base/twiceprecision.jl

```diff
@@ -0,0 +1,503 @@
+# Note this returns a StepRangeLen
+function linspace{T}(::Type{T}, start_n::Integer, stop_n::Integer, len::Integer, den::Integer)
+    len < 2 && return _linspace1(T, start_n/den, stop_n/den, len)
+    start_n == stop_n && return StepRangeLen(TwicePrecision{T}((start_n, den)), zero(TwicePrecision{T}), len)
+    tmin = -start_n/(Float64(stop_n) - Float64(start_n))
+    imin = round(Int, tmin*(len-1)+1)
+    imin = clamp(imin, 1, Int(len))
+    # Compute (1-t)*a and t*b separately in 2x precision (itp = interpolant)...
+    dent = (den, len-1)  # represent products as a tuple to eliminate risk of overflow
+    start_itp = proddiv(T, (len-imin, start_n), dent)
+    stop_itp = proddiv(T, (imin-1, stop_n), dent)
+    # ...and then combine them to make ref
+    ref = start_itp + stop_itp
+    # Compute step to 2x precision without risking overflow...
+    rend = proddiv(T, (stop_n,), dent)
+    rbeg = proddiv(T, (-start_n,), dent)
+    step = twiceprecision(rbeg + rend, nbitslen(len, imin)) # ...and truncate hi-bits as needed
+    StepRangeLen(ref, step, Int(len), imin)
+end
+
+# For len < 2
+function _linspace1{T}(::Type{T}, start, stop, len)
+    len >= 0 || throw(ArgumentError("linspace($start, $stop, $len): negative length"))
+    if len <= 1
+        len == 1 && (start == stop || throw(ArgumentError("linspace($start, $stop, $len): endpoints differ")))
+        # Ensure that first(r)==start and last(r)==stop even for len==0
+        return StepRangeLen(TwicePrecision(start, zero(T)), TwicePrecision(start, -stop), len, 1)
+    end
+    throw(ArgumentError("should only be called for len < 2, got $len"))
+end
+
+### Numeric utilities
+
+# approximate x with a rational representation. Guaranteed to return,
+# but not guaranteed to return a precise answer.
+function rat(x)
+    y = x
+    a = d = 1
+    b = c = 0
+    m = maxintfloat(narrow(typeof(x)))
+    while abs(y) <= m
+        f = trunc(Int,y)
+        y -= f
+        a, c = f*a + c, a
+        b, d = f*b + d, b
+        max(abs(a), abs(b)) <= convert(Int,m) || return c, d
+        oftype(x,a)/oftype(x,b) == x && break
+        y = inv(y)
+    end
+    return a, b
+end
+
+narrow(::Type{Float64}) = Float32
+narrow(::Type{Float32}) = Float16
+narrow(::Type{Float16}) = Float16
+
+function add2{T<:Number}(u::T, v::T)
+    @_inline_meta
+    u, v = ifelse(abs(v) > abs(u), (v, u), (u, v))
+    w = u + v
+    w, (u-w) + v
+end
+
+add2(u, v) = _add2(promote(u, v)...)
+_add2{T<:Number}(u::T, v::T) = add2(u, v)
+_add2(u, v) = error("$u::$(typeof(u)) and $v::$(typeof(v)) cannot be promoted to a common type")
+
+function +(x::TwicePrecision, y::Number)
+    s_hi, s_lo = add2(x.hi, y)
+    TwicePrecision(s_hi, s_lo+x.lo)
+end
++(x::Number, y::TwicePrecision) = y+x
+
+function +{T}(x::TwicePrecision{T}, y::TwicePrecision{T})
+    r = x.hi + y.hi
+    s = abs(x.hi) > abs(y.hi) ? (((x.hi - r) + y.hi) + y.lo) + x.lo : (((y.hi - r) + x.hi) + x.lo) + y.lo
```

Comment:
I'm almost certain it is, but @simonbyrne is the expert here. (That code comes straight from his DoubleDouble.jl.)